mailpoet/lib/Automation/Integrations/WooCommerce/Fields/CustomerOrderFieldsFactory.php

```diff
@@ -0,0 +1,217 @@
+        'woocommerce:customer:last-paid-order-date',
+        Field::TYPE_DATETIME,
+        __('Last paid order date', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          if (!$customer) {
+            return null;
+          }
+          return $this->getPaidOrderDate($customer, false);
+        }
+      ),
+      new Field(
+        'woocommerce:customer:purchased-categories',
+        Field::TYPE_ENUM_ARRAY,
+        __('Purchased categories', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          return $customer ? $this->getOrderProductTermIds($customer, 'product_cat') : [];
+        },
+        [
+          'options' => $this->getTermsList('product_cat'),
+        ]
+      ),
+      new Field(
+        'woocommerce:customer:purchased-tags',
+        Field::TYPE_ENUM_ARRAY,
+        __('Purchased tags', 'mailpoet'),
+        function (CustomerPayload $payload) {
+          $customer = $payload->getCustomer();
+          return $customer ? $this->getOrderProductTermIds($customer, 'product_tag') : [];
+        },
+        [
+          'options' => $this->getTermsList('product_tag'),
+        ]
+      ),
+    ];
+  }
+
+  private function getPaidOrderDate(WC_Customer $customer, bool $fetchFirst): ?DateTimeImmutable {
+    $wpdb = $this->wordPress->getWpdb();
+    $sorting = $fetchFirst ? 'ASC' : 'DESC';
+    $statuses = array_map(function (string $status) {
+      return "wc-$status";
+    }, $this->wooCommerce->wcGetIsPaidStatuses());
+    $statusesPlaceholder = implode(',', array_fill(0, count($statuses), '%s'));
+
+    if ($this->wooCommerce->isWooCommerceCustomOrdersTableEnabled()) {
+      $statement = (string)$wpdb->prepare("
+        SELECT o.date_created_gmt
+        FROM {$wpdb->prefix}wc_orders o
+        WHERE o.customer_id = %d
+        AND o.status IN ($statusesPlaceholder)
+        AND o.total_amount > 0
+        ORDER BY o.date_created_gmt {$sorting}
+        LIMIT 1
+      ", array_merge([$customer->get_id()], $statuses));
+    } else {
+      $statement = (string)$wpdb->prepare("
+        SELECT p.post_date_gmt
+        FROM {$wpdb->prefix}posts p
+        LEFT JOIN {$wpdb->prefix}postmeta pm_total ON p.ID = pm_total.post_id AND pm_total.meta_key = '_order_total'
+        LEFT JOIN {$wpdb->prefix}postmeta pm_user ON p.ID = pm_user.post_id AND pm_user.meta_key = '_customer_user'
+        WHERE p.post_type = 'shop_order'
+        AND p.post_status IN ($statusesPlaceholder)
+        AND pm_user.meta_value = %d
+        AND pm_total.meta_value > 0
+        ORDER BY p.post_date_gmt {$sorting}
+        LIMIT 1
+      ", array_merge($statuses, [$customer->get_id()]));
+    }
+
+    $date = $wpdb->get_var($statement);
+    return $date ? new DateTimeImmutable($date, new DateTimeZone('GMT')) : null;
+  }
+
+  private function getOrderProductTermIds(WC_Customer $customer, string $taxonomy): array {
```

Comment:
Here again, I am not sure, if we should directly query the database. There are reasons to do so sometimes (performance), but not using the WooCommerce and WP Api comes at a cost: Skipping several caching layers and ignoring the `applyFilters()` which might integrate custom logic.

Tbh at the moment I am not sure which side I lean towards

Reply:
@websupporter Similarly as in the previous case, we'd need to iterate all orders of the customer (paginated), then iterate all products withing them, and then fetch all tag/category IDs from those products. That is a huge amount of data and CPU time to fetch just an array of IDs. At some point HPOS will be the default, and the complex non-HPOS branches will be deleted.